North-Seattle-College/ad440-winter2022-tuesday-repo
# React s3 deploy

# Documentation:

Closes #8 

Updates PR #34 
- UPDATE:  This PR now includes a `github/workflows` folder system with a working `.yml` file that  when pushed to development will run a `build` of the React App in a S3 bucket and run the App via AWS' static Website hosting.

- This PR updated the documentation based on research and testing on how to deploy React to an S3 bucket utilizing GitHub Actions
- This includes a frontend documentation folder to store these instructions as well as future issues on the frontend side that require a document/reference.
- The file instructions include:
   -  Creating an S3 Bucket _(while not directly part of this issue, were included)_
   -  Setting up the S3 Bucket for Web Hosting _(again not directly part of this issue, but included)_
   -  Configuring GitHub Actions workflow to deploy changes automatically. _(these have also been included below since this is the subject of this PR)_

### __(User Story, Total time to complete the Sprint, and Acceptance Criteria can be found on the connected Issue)__

## File Locations
The `S3-deploy.yml` file can be found in the `.github/workfolder` folder for reference
The `reactS3Action.md` file can be found in the `frontend/documentation` folder for reference.

## Time Spent on Issue:
| Task Breakdown Description | Approx. Time |
| ------------- | ------------- |
| Troubleshooting the .yml file to run successfully | 4 hrs |
| Researching Deployment Resources | 2 hrs |
| Setting up React  Environment| 1 hr  |
| Compile the References to one easy to follow instruction  | 0.5 hr  |
| Testing Part 1: Create a test Repo, and React App | 1 hr  |
| Testing Part 2: Create a test S3 Bucket | 0.5 hr  |
| Testing Part 3:  Setup the Github Secrets and Workflow | 2 hrs | 
| Create the .md file and documentation folder in the frontend folder | 2 hrs |
| Create the PR with additional information and follow-along screenshots | 1 hr |
| Frontend Team Meets (as of 1/24/2022) | 3 hrs (total of all meetings)|
| Collaborate with DevOps Team | 1 hr |

# Gallery & GitHub Action specific Instruction:

## On the GitHub Repo:
1. Click on the __Settings__ Tab.
2. Select __Secret__ on the left side menu
3. Click on __New Secret__ to add a secret providing _Name_ and _Value_:

 
![Secrets](https://user-images.githubusercontent.com/52547964/150889344-95acd96e-d707-432f-938c-9f0ed3087cc3.PNG)

______________________

### Create key/value pairs for the following sensitive information:

| Name:          | Value:         |
|:---------------|:---------------|
|AWS_ACCESS_KEY_ID| your-aws-acess-key-id |
|AWS_SECRET_ACCESS_KEY| your-aws-secret-access-key |
|AWS_REGION| us-west-2 or other region |
|AWS_PRODUCTION_BUCKET_NAME| your-aws-bucket|

_____________________

### Be sure to note the names of the keys:
![GitHubActions](https://user-images.githubusercontent.com/52547964/150889991-d6a22ef1-47f6-4f63-92ac-0fd8cde0bcf9.PNG)

______________________


### How the secrets will show once submitted:
![RepoSecrets](https://user-images.githubusercontent.com/52547964/150889648-afa517b7-af57-4c98-8765-9693b168e8a2.PNG)

______________________

### Setup GitHub Actions
1. While In the Github repo, click on the __Actions__ tab.
2. On the __Actions__ page, click on the "__Set up this workflow__" or __Set up a workflow yourself__ button. _This takes you to a new page with a web editor with some container code which can be copied and deleted_
3. Copy the container code it generates for the next step.
![GitHubActions2](https://user-images.githubusercontent.com/52547964/150890771-3c66434a-2b6f-477c-a4fe-f7f28e700a7a.PNG)

_______________________

### Return to your working Branch
1. Create a `.github` folder
2. In the `.github` folder create another folder named `workflows`
3. while in the `workflows` create a `.yml` file with a name that describes the action (example `S3-deploy.yml`)
![1 File Structure](https://user-images.githubusercontent.com/52547964/152284212-09aef168-2676-4de7-8cd1-82543ed02cc8.PNG)

_________________________

4. Paste the filler `.yml` provided by GitHub
5. Make the following adjustments to the `.yml` file:
![yml](https://user-images.githubusercontent.com/52547964/152284377-b2fb3366-5920-4374-9a2f-115aa2652126.PNG)

_______________________

### What this file obtains:
The file is made of 3 different parts to help with the understanding:
* __name:__  The name of the Workflow
* __on:__ what triggers the Workflow, in this case the workflow will run anytime you __push__ code to the required __development branch__. _(It will trigger again when pushed to __master__ but no code will be directly pushed to __master__ in this specific project)_
* __jobs:__ Workflow run making up one or more jobs and they run parallel by default:
    * __steps:__ A job contains a sequence of tasks.. Steps run commands nad actions in your repo.
    * __actions/checkout@v2:__ Checks out the repo so workflow can access it.
    * __aws-actions/configure-aws-credentials@v1:__ This configures AWS credentials & region environment for use in other GitHub Actions
    * __Build React App:__ Installs node packages and runs the build in the __package.json__ file, and creates a __build__ folder in the root directory.
    * __Deploy app build to S3 bucket:__  This deploys the newly created build to S3 bucket _bucket-name_ (replace with the name of the created S3 bucket)

To save, click on the __Start Commit__ then __Commit New File__.
Doing so will:
* Save the Action creating a .github directory with a workflows directory in it.
* It will contained the file s3-deploy (or whatever name assigned to it)
* Trigger the actions assigned.

![WorkflowAdded](https://user-images.githubusercontent.com/52547964/150891153-225fb560-dc9e-40b8-bf29-04e887a201b4.PNG)

______________________________

### This should successfully complete the setup. 

### Running the Build
* Once you push the working branch you should run a successful build
![build](https://user-images.githubusercontent.com/52547964/152285628-956cd55c-9488-4078-91cb-79f73b5b4a93.PNG)

_____________________________

### Test the AWS Website
1.  Go to the bucket where the app was deployed and make your the files contained in the build folder carried over.
![Bucket](https://user-images.githubusercontent.com/52547964/152285808-759fa6da-aca5-4db5-b964-f519708ed8bc.PNG)

_____________________________
2. While on the bucket, click on the `Properties` tab scroll to the bottom to __Static Website Hosting__ to find the bucket website endpoint.
![StaticHosting](https://user-images.githubusercontent.com/52547964/152286045-2fc33ccc-dc12-4882-93fd-5cc5eaf1f2f0.PNG)

_____________________________
3. Click the link and it should display the working website!  __Congrats!!__
![WebSite](https://user-images.githubusercontent.com/52547964/152286187-88d69cbe-e4ae-4ba8-a7de-bb55cdeaeb71.PNG)


_____________________________

## Troubleshooting:
*  Certain IAM permissions may be needed on the creation of the S3 bucket, utilizing the correct S3 bucket creation format
* As mentioned in the documentation, setting up the .yaml build file may require some additional steps in the jobs portion than what was added to the instruction such as branch names for pushing to and PR requests.
_(See documentation file for more details)_
